atomic_reactor/plugins/pre_pull_base_image.py

```diff
@@ -65,92 +87,78 @@ def resolve_base_image(self, build_json):
+                image, self.parent_registry
+            )
+            image_with_registry.registry = self.parent_registry
         else:
-            self.log.info("pulling base image '%s'", base_image)
-
-        if self.check_platforms:
-            self.validate_platforms_in_base_image(base_image_with_registry)
+            self.log.info("pulling parent image '%s'", image)
 
-        try:
-            self.tasker.pull_image(base_image_with_registry,
-                                   insecure=self.parent_registry_insecure)
-
-        except RetryGeneratorException as original_exc:
-            if base_image_with_registry.namespace:
-                raise
-
-            self.log.info("'%s' not found", base_image_with_registry.to_str())
-            base_image_with_registry.namespace = 'library'
-            self.log.info("trying '%s'", base_image_with_registry.to_str())
+        return image_with_registry
 
+    def _pull_and_tag_image(self, image, build_json):
+        """Docker pull the image and tag it uniquely for use by this build"""
+        image = image.copy()
+        first_library_exc = None
+        for _ in range(20):
             try:
-                self.tasker.pull_image(base_image_with_registry,
-                                       insecure=self.parent_registry_insecure)
+                self.tasker.pull_image(image, insecure=self.parent_registry_insecure)
+                self.workflow.pulled_base_images.add(image.to_str())
+            except RetryGeneratorException as exc:
+                # the pull itself failed
+                if first_library_exc:
+                    raise first_library_exc
+                if image.namespace:
+                    raise
+
+                self.log.info("'%s' not found", image.to_str())
+                image.namespace = 'library'
+                self.log.info("trying '%s'", image.to_str())
+                first_library_exc = exc
+                continue
+
+            # Attempt to tag it using a unique ID. We might have to retry
+            # if another build with the same parent image is finishing up
+            # and removing images it pulled.
+
+            # Use the OpenShift build name as the unique ID
+            unique_id = build_json['metadata']['name']
+            new_image = ImageName(repo=unique_id)
 
-            except RetryGeneratorException:
-                raise original_exc
-
-        pulled_base = base_image_with_registry.to_str()
-        self.workflow.pulled_base_images.add(pulled_base)
-
-        # Attempt to tag it using a unique ID. We might have to retry
-        # if another build with the same parent image is finishing up
-        # and removing images it pulled.
-
-        # Use the OpenShift build name as the unique ID
-        unique_id = build_json['metadata']['name']
-        original_base_image = base_image.copy()
-        base_image = ImageName(repo=unique_id)
-
-        for _ in range(20):
```

Comment:
any reason why we try this 20 times, as opposed to 10 or 5 or 100? I get that you're inheriting this code, mostly curious.

Reply:
I don't know, of course. You only go through the loop if the image pulls successfully but gets immediately deleted before being tagged. That ought to be an edge case race condition, so 20 seems excessive, but on the other hand there could be a *lot* of builds running with the same parents, and retries are cheap.

Reply:
Yes, thanks.